### PR TITLE
Rework docs for the `size` of `terms` agg

### LINKED ISSUE #72684: Slightly more accurate `terms` sorting on sub-aggs
Right now we default the `shard_size` to `1.5*size+10` in `terms` in an effort to keep the doc count errors lower. When you sort the terms by something other than doc_count descending the error is unbounded. But larger `shard_size` will lead to more accurate results because we ship more data back to the coordinating node. Maybe we should bump the `shard_size` if you are sorting on anything other than doc_count descending to get you more accurate results by default. Sure, it'll be slower, but it's the price you pay for a little more accuracy. And the performance when sorting by the default won't change.

### PR DESCRIPTION
**Pull request by nik9000**
The `terms` agg picks the top `size` terms in a single scatter/gather
pass across all the shards. For the default `order` and if you `order`
by `_key` this works quite well. Some errors creep in, but it's fairly
easy to point to them and understand them. But ordering by doc count
ascending is like inviting the error vampire into your agg. It's super
easy to get inaccurate results. This updates the docs to be more stark
about it.

Closes #72684

### CODE REVIEW

**Review by jrodewig — CHANGES_REQUESTED** [review]
Thanks for putting this together, @nik9000.

For this first pass, I focused mostly on the `order` warnings. However, I also left some feedback on the other sections. Don't feel obligated to take all the suggestions.

I'd like to take another look once you feel this is ready for it.

**Review by jrodewig — APPROVED** [review]
Thanks for iterating on this @nik9000. Overall, this LGTM.

I mostly left minor style nits that you can disregard if wanted. The only bigger piece of feedback is in the `size` section. We may want to provide clearer guidance on:

- How many terms are "too many" when choosing between the terms agg and the composite agg.
- What happens if you try using a huge `size` on a high-cardinality field.

**Comment by nik9000** [comment]
run elasticsearch-ci/part-2

**Comment by nik9000** [comment]
OK! I'll backport it myself.

File: server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java [hunk 1]
```diff
@@ -16,29 +16,29 @@
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.mapper.DateFieldMapper;
 import org.elasticsearch.index.shard.IndexLongFieldRange;
-import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:19]: Oh yikes this snuck in here! That was not my intent....

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:19]: It's supposed to be in https://github.com/elastic/elasticsearch/pull/80161 - which was approved. So I'm not going to revert.

Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/query/CoordinatorRewriteContextProvider.java:19]: But I won't backport it.

**Comment by nik9000** [comment]
Backported in #80223, #80226, and #80227.

**Pull request merged**